handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -1501,16 +1502,21 @@ private void setHandshakeFailure(ChannelHandlerContext ctx, Throwable cause, boo
                     }
                 }
             }
-            notifyHandshakeFailure(cause);
+            notifyHandshakeFailure(cause, notify);
         } finally {
             // Ensure we remove and fail all pending writes in all cases and so release memory quickly.
             pendingUnencryptedWrites.releaseAndFailAll(ctx, cause);
         }
     }
 
-    private void notifyHandshakeFailure(Throwable cause) {
+    private void notifyHandshakeFailure(Throwable cause, boolean notify) {
         if (handshakePromise.tryFailure(cause)) {
-            SslUtils.notifyHandshakeFailure(ctx, cause);
+            if (notify) {
+                SslUtils.notifyHandshakeFailure(ctx, cause);
```

Comment:
would it make sense to move the conditional logic into `SslUtils.notifyHandshakeFailure`? SniHandler may fire an exception if it sees something that is not an SSL record, and the handshake may not have started.

Reply:
I guess we could.